homeassistant/components/fan/reproduce_state.py

```diff
@@ -32,14 +32,23 @@
 _LOGGER = logging.getLogger(__name__)
 
 VALID_STATES = {STATE_ON, STATE_OFF}
-ATTRIBUTES = {  # attribute: service
+
+SIMPLE_ATTRIBUTES = {  # attribute: service
     ATTR_DIRECTION: SERVICE_SET_DIRECTION,
     ATTR_OSCILLATING: SERVICE_OSCILLATE,
+}
+
+# These need to be handled separately because they are mutually exclusive.
+# The order of these is important, as SPEED is legacy and should be checked for
+# first.
+SPEED_AND_MODES_ATTRIBUTES = {
     ATTR_SPEED: SERVICE_SET_SPEED,
     ATTR_PERCENTAGE: SERVICE_SET_PERCENTAGE,
     ATTR_PRESET_MODE: SERVICE_SET_PRESET_MODE,
 }
 
+ATTRIBUTES = set(SIMPLE_ATTRIBUTES.keys()) | set(SPEED_AND_MODES_ATTRIBUTES)
```

Comment:
I think this can be removed if we remove the  `Return if we are already at the right state.` check at the top. It seems like the code below should be a no-op if all the attributes are the same so i'm not sure its worth preserving given the extra code being added, and lets us remove this variable.

Reply:
There was one small change needed to maintain behaviour. I've fixed that now.